### ISSUE TITLE
Driver loading twice on 7700x

### ISSUE THREAD

**Issue by GibsonSoft** [amkillam/ryzen_smu#4]
Hey there, I kind of hit a brick wall with this and I'm not sure what to do next.

It looks like the driver is attempting to load twice on my Proxmox installation. At least, that's what it looks like from dmesg:
```
root@pve:~# dmesg | grep smu
[    4.884998] ryzen_smu: loading version: 0.1.5
[    4.885002] ryzen_smu: CPUID: family 0x19, model 0x61, stepping 0x2, package 0x0
[    4.885004] ryzen_smu: RSMU Mailbox: (cmd: 0x3B10524, rsp: 0x3B10570, args: 0x3B10A40)
[    4.885007] ryzen_smu: HSMP Mailbox: (cmd: 0x3B10534, rsp: 0x3B10980, args: 0x3B109E0)
[    4.885008] ryzen_smu: MP1 Mailbox: (cmd: 0x3B10530, rsp: 0x3B1057C, args: 0x3B109C4)
[    4.885010] ryzen_smu: Family Codename: Raphael
[    4.885011] ryzen_smu: SMU Service Request: ID(0x2) Args(0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
[    4.885047] ryzen_smu: SMU Service Response: ID(0x2) Args(0x544fe1, 0x0, 0x0, 0x0, 0x0, 0x0)
[    4.885050] ryzen_smu: SMU v84.79.225
[    4.885051] ryzen_smu: SMU Service Request: ID(0x2) Args(0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
[    4.885104] ryzen_smu: SMU Service Response: ID(0x2) Args(0x544fe1, 0x0, 0x0, 0x0, 0x0, 0x0)
[    4.885108] ryzen_smu: SMU Service Request: ID(0x3) Args(0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
[    4.885158] ryzen_smu: SMU Service Response: ID(0x3) Args(0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
[    4.885159] ryzen_smu: SMU Service Request: ID(0x5) Args(0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
[    4.885303] ryzen_smu: SMU Service Response: ID(0x5) Args(0x540104, 0x0, 0x0, 0x0, 0x0, 0x0)
[    4.885308] ryzen_smu: Probing the PM table for state changes
[    4.885310] ryzen_smu: SMU Service Request: ID(0x4) Args(0x1, 0x1, 0x0, 0x0, 0x0, 0x0)
[    4.885606] ryzen_smu: SMU Service Response: ID(0x4) Args(0x3ef00000, 0x10, 0x0, 0x0, 0x0, 0x0)
[    4.885609] ryzen_smu: SMU Service Request: ID(0x5) Args(0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
[    4.885640] ryzen_smu: SMU Service Response: ID(0x5) Args(0x540104, 0x0, 0x0, 0x0, 0x0, 0x0)
[    4.885642] ryzen_smu: Determined PM mapping size as (6a8h,0h) bytes.
[    4.885644] ryzen_smu: SMU Service Request: ID(0x3) Args(0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
[    4.885775] ryzen_smu: SMU Service Response: ID(0x3) Args(0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
[    4.885783] ryzen_smu: Probe succeeded: read 1704 bytes
[    4.885816] ryzen_smu: loading version: 0.1.5
[    4.885817] ryzen_smu: SMU Service Request: ID(0x2) Args(0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
[    4.885825] ryzen_smu: SMU Service Response: ID(0x2) Args(0x14e01022, 0x14e01022, 0x14e01022, 0x14e01022, 0x14e01022, 0x14e01022)
[    4.885828] ryzen_smu: SMU v20.224.16.34
[    4.885829] ryzen_smu: SMU Service Request: ID(0x2) Args(0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
[    4.885837] ryzen_smu: SMU Service Response: ID(0x2) Args(0x14e01022, 0x14e01022, 0x14e01022, 0x14e01022, 0x14e01022, 0x14e01022)
[    4.885840] ryzen_smu: SMU Service Request: ID(0x3) Args(0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
[    4.885848] ryzen_smu: SMU Service Response: ID(0x3) Args(0x14e01022, 0x14e01022, 0x14e01022, 0x14e01022, 0x14e01022, 0x14e01022)
[    4.885851] ryzen_smu: SMU Service Request: ID(0x5) Args(0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
[    4.885859] ryzen_smu: SMU Service Response: ID(0x5) Args(0x14e01022, 0x14e01022, 0x14e01022, 0x14e01022, 0x14e01022, 0x14e01022)
[    4.885861] ryzen_smu: Probing the PM table for state changes
[    4.885863] ryzen_smu: Probe succeeded: read 1704 bytes
[    4.885866] sysfs: cannot create duplicate filename '/kernel/ryzen_smu_drv'
[    4.885907]  ryzen_smu_probe+0x118/0x370 [ryzen_smu]
[    4.885949]  ? __pfx_ryzen_smu_driver_init+0x10/0x10 [ryzen_smu]
[    4.885955]  ryzen_smu_driver_init+0x23/0xff0 [ryzen_smu]
[    4.886080] kobject: kobject_add_internal failed for ryzen_smu_drv with -EEXIST, don't try to register things with the same name in the same directory.
[    4.886084] ryzen_smu: Unable to create sysfs interface
[    4.886086] ryzen_smu 0000:00:18.4: probe with driver ryzen_smu failed with error -12
```

It does appear to be sending info to `/sys/kernel/ryzen_smu_drv`:
```
root@pve:~# cat /sys/kernel/ryzen_smu_drv/drv_version
0.1.5
```

```
root@pve:~# cat /sys/kernel/ryzen_smu_drv/version
20.224.16.34
```

```
root@pve:~# cat /sys/kernel/ryzen_smu_drv/mp1_if_version 
2
```

```
root@pve:~# cat /sys/kernel/ryzen_smu_drv/codename 
20
```

Additional stack trace if it's helpful:
```
[    4.885875] Call Trace:
[    4.885877]  <TASK>
[    4.885880]  dump_stack_lvl+0x76/0xa0
[    4.885885]  dump_stack+0x10/0x20
[    4.885887]  sysfs_warn_dup+0x8a/0xb0
[    4.885892]  sysfs_create_dir_ns+0xe0/0x100
[    4.885896]  kobject_add_internal+0xa8/0x2f0
[    4.885899]  kobject_add+0x93/0x100
[    4.885904]  kobject_create_and_add+0x85/0xe0
[    4.885907]  ryzen_smu_probe+0x118/0x370 [ryzen_smu]
[    4.885912]  local_pci_probe+0x44/0xb0
[    4.885917]  pci_device_probe+0xc5/0x260
[    4.885921]  really_probe+0xee/0x3b0
[    4.885925]  __driver_probe_device+0x8c/0x180
[    4.885929]  driver_probe_device+0x24/0xd0
[    4.885932]  __driver_attach+0x10b/0x210
[    4.885934]  ? __pfx___driver_attach+0x10/0x10
[    4.885937]  bus_for_each_dev+0x8a/0xf0
[    4.885941]  driver_attach+0x1e/0x30
[    4.885943]  bus_add_driver+0x14e/0x290
[    4.885947]  driver_register+0x5e/0x130
[    4.885949]  ? __pfx_ryzen_smu_driver_init+0x10/0x10 [ryzen_smu]
[    4.885953]  __pci_register_driver+0x5e/0x70
[    4.885955]  ryzen_smu_driver_init+0x23/0xff0 [ryzen_smu]
[    4.885958]  do_one_initcall+0x5b/0x340
[    4.885963]  do_init_module+0x97/0x2a0
[    4.885967]  load_module+0x1eb0/0x2170
[    4.885974]  init_module_from_file+0x96/0x100
[    4.885977]  ? srso_alias_return_thunk+0x5/0xfbef5
[    4.885980]  ? init_module_from_file+0x96/0x100
[    4.885985]  idempotent_init_module+0x11c/0x310
[    4.885989]  __x64_sys_finit_module+0x64/0xd0
[    4.885992]  x64_sys_call+0x1697/0x24e0
[    4.885995]  do_syscall_64+0x7e/0x170
[    4.885998]  ? srso_alias_return_thunk+0x5/0xfbef5
[    4.886000]  ? sysvec_call_function+0x57/0xc0
[    4.886003]  ? srso_alias_return_thunk+0x5/0xfbef5
[    4.886005]  ? asm_sysvec_call_function+0x1b/0x20
[    4.886010]  ? __entry_text_end+0x10258d/0x10258d
[    4.886012]  ? srso_alias_return_thunk+0x5/0xfbef5
[    4.886014]  ? syscall_exit_to_user_mode+0x4e/0x250
[    4.886016]  ? srso_alias_return_thunk+0x5/0xfbef5
[    4.886018]  ? do_syscall_64+0x8a/0x170
[    4.886022]  ? srso_alias_return_thunk+0x5/0xfbef5
[    4.886024]  ? __do_sys_newfstatat+0x53/0x90
[    4.886030]  ? srso_alias_return_thunk+0x5/0xfbef5
[    4.886032]  ? syscall_exit_to_user_mode+0x4e/0x250
[    4.886034]  ? srso_alias_return_thunk+0x5/0xfbef5
[    4.886036]  ? do_syscall_64+0x8a/0x170
[    4.886038]  ? syscall_exit_to_user_mode+0x4e/0x250
[    4.886040]  ? srso_alias_return_thunk+0x5/0xfbef5
[    4.886042]  ? do_syscall_64+0x8a/0x170
[    4.886044]  ? srso_alias_return_thunk+0x5/0xfbef5
[    4.886046]  ? switch_fpu_return+0x50/0xe0
[    4.886050]  ? srso_alias_return_thunk+0x5/0xfbef5
[    4.886052]  ? syscall_exit_to_user_mode+0x18d/0x250
[    4.886054]  ? srso_alias_return_thunk+0x5/0xfbef5
[    4.886056]  ? do_syscall_64+0x8a/0x170
[    4.886058]  entry_SYSCALL_64_after_hwframe+0x76/0x7e
[    4.886061] RIP: 0033:0x7fa2014d1799
[    4.886064] Code: 08 89 e8 5b 5d c3 66 2e 0f 1f 84 00 00 00 00 00 90 48 89 f8 48 89 f7 48 89 d6 48 89 ca 4d 89 c2 4d 89 c8 4c 8b 4c 24 08 0f 05 <48> 3d 01 f0 ff ff 73 01 c3 48 8b 0d 37 06 0d 00 f7 d8 64 89 01 48
[    4.886066] RSP: 002b:00007fff1f41c978 EFLAGS: 00000246 ORIG_RAX: 0000000000000139
[    4.886069] RAX: ffffffffffffffda RBX: 000062a03c5917d0 RCX: 00007fa2014d1799
[    4.886070] RDX: 0000000000000000 RSI: 00007fa201664efd RDI: 0000000000000006
[    4.886072] RBP: 00007fa201664efd R08: 0000000000000000 R09: 000062a03c552ee0
[    4.886073] R10: 0000000000000006 R11: 0000000000000246 R12: 0000000000020000
[    4.886074] R13: 0000000000000000 R14: 000062a03c5d53c0 R15: 00007fff1f41cbb0
[    4.886078]  </TASK>
```

**Comment by GibsonSoft** [comment]
> It seems like the probe function is getting called for two devices present in the system. Now, the question is, which ones?

That was my thought too. Only thing I can think of is it's being called for both the CPU and iGPU. I'm not sure if that even makes sense, though 🤷‍♂️ I was browsing the code when I originally posted the issue, but it's been so long I totally forgot where I left off lol

**Comment by tedliosu** [comment]
Hello @eaglgenes101 and @GibsonSoft; I am currently running into this same issue on my 7700x system as well.

> It seems like the probe function is getting called for two devices present in the system. Now, the question is, which ones?

On my machine, if I add the following line of code to the very beginning of the `ryzen_smu_probe` function in drv.c:

```c
pr_info("---INFORMATION - loading for device vendor 0x%x --- device 0x%x ---", id->vendor, id->device);
```

I get the following output in my `sudo -H dmesg`:

```
[  373.580764] ryzen_smu: ---INFORMATION - loading for device vendor 0x1022 --- device 0x14d8 ---
[  373.580766] ryzen_smu: loading version: 0.1.5
[  373.580768] ryzen_smu: CPUID: family 0x19, model 0x61, stepping 0x2, package 0x0
[  373.580769] ryzen_smu: Family Codename: Raphael
[  373.580801] ryzen_smu: SMU v84.79.225
[  373.581058] ryzen_smu: ---INFORMATION - loading for device vendor 0x1022 --- device 0x14e4 ---
[  373.581059] ryzen_smu: loading version: 0.1.5
[  373.581067] ryzen_smu: SMU v20.224.16.34
[  373.581097] sysfs: cannot create duplicate filename '/kernel/ryzen_smu_drv'
```

Given that (at least on my Ubuntu Linux machine) there exists a `/usr/share/misc/pci.ids` file that helps us to interpret these hex device vendor and device codes, I pulled it up and scrolled down to the "1022  Advanced Micro Devices, Inc. [AMD]" section and found the following:

>         14d8  Raphael/Granite Ridge Root Complex
>         14d9  Raphael/Granite Ridge IOMMU
>         14da  Raphael/Granite Ridge Dummy Host Bridge
>         14db  Raphael/Granite Ridge GPP Bridge
>         14dc  SDXI
>         14dd  Raphael/Granite Ridge Internal GPP Bridge to Bus [C:A]
>         14de  Raphael/Granite Ridge PCIe Dummy Function
>         14e0  Raphael/Granite Ridge Data Fabric; Function 0
>         14e1  Raphael/Granite Ridge Data Fabric; Function 1
>         14e2  Raphael/Granite Ridge Data Fabric; Function 2
>         14e3  Raphael/Granite Ridge Data Fabric; Function 3
>         14e4  Raphael/Granite Ridge Data Fabric; Function 4
>         14e5  Raphael/Granite Ridge Data Fabric; Function 5
>         14e6  Raphael/Granite Ridge Data Fabric; Function 6
>         14e7  Raphael/Granite Ridge Data Fabric; Function 7

So basically to answer your question of which devices are this `ryzen_smu` driver being initialized for, it seems to be for both the Root Complex and one of the Data Fabric Functions, the former which makes sense but the latter I have no idea what it is since unfortunately I'm not familiar with the internals of an AMD CPU 🙃

But hopefully this helps to answer y'all's question 👍
